core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  @Override
+  public void cleanFiles() {
+    // Identify all of the manifest lists to retain
+    Set<String> manifestListsBeforeExpiration =
+        ReachableFileUtil.manifestListLocations(beforeExpiration);
+    Map<String, ManifestFile> manifestsBeforeExpiration =
+        computeManifestsForManifestLists(manifestListsBeforeExpiration);
+    // Identify all of the manifest lists that currently exist
+    Set<String> manifestListsAfterExpiration =
+        ReachableFileUtil.manifestListLocations(afterExpiration);
+    Map<String, ManifestFile> manifestsAfterExpiration =
+        computeManifestsForManifestLists(manifestListsAfterExpiration);
+
+    // The manifest files which we could delete are just the ones which existed before expiration
+    // which no longer exist
+    Set<String> manifestListsToDelete =
+        Sets.difference(manifestListsBeforeExpiration, manifestListsAfterExpiration);
+
+    MapDifference<String, ManifestFile> diff =
+        Maps.difference(manifestsBeforeExpiration, manifestsAfterExpiration);
+
+    List<ManifestFile> manifestsToDelete = Lists.newArrayList(diff.entriesOnlyOnLeft().values());
+    List<ManifestFile> currentManifests = Lists.newArrayList(manifestsAfterExpiration.values());
+
+    Set<String> filesToDelete = findFilesToDelete(manifestsToDelete, currentManifests);
+    Tasks.foreach(filesToDelete)
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Delete failed for data file: {}", file, exc))
+        .run(deleteFunc::accept);
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    deleteMetadataFiles(manifestPathsToDelete, manifestListsToDelete);
+  }
+
+  private Map<String, ManifestFile> computeManifestsForManifestLists(Set<String> manifestLists) {
+    Map<String, ManifestFile> pathToManifestFiles = Maps.newHashMap();
+    for (String manifestList : manifestLists) {
+      CloseableIterable<ManifestFile> manifestFiles = readManifestFiles(manifestList);
+      for (ManifestFile manifest : manifestFiles) {
+        pathToManifestFiles.put(manifest.path(), manifest.copy());
+      }
+    }
+    return pathToManifestFiles;
+  }
+
+  private void deleteMetadataFiles(
+      Set<String> manifestsToDelete, Set<String> manifestListsToDelete) {
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+
+    Tasks.foreach(manifestsToDelete)
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((manifest, exc) -> LOG.warn("Delete failed for manifest: {}", manifest, exc))
+        .run(deleteFunc::accept);
```

Comment:
Done, although I think down the line in a follow on PR , we should update so this uses batch delete if applicable. This may change the structure a bit at that point in time. But it makes sense to do at this point in time, thanks!